#130 redux refactor

bkmartinjr opened the pull request:
This resolves almost all of #130

Redux refactor. This PR includes:

- added a `utils/stateManager` module which provides an abstraction layer
  between the REST API and components. Features:
  - Clean support for cases where world != universe (eg, Regraph UX)
  - Separates OTA formats from in-app formats - transforms on load
  - caching of expression (var) data
- restructured all reducers to use the new universe/world state manager:
  - removed: initialize, expression, cells
  - controls rewritten and renamed controls2
- rewrote most of the action creators to use the new reducer model. Several
  also converted from promises to async/await.
- converted several Component data caches to use memoization, per React best practices.  This required adding memoize-one package.
- removed a number of dead modules (joy plots, parallel, etc)
- Babel update to use modern-browser profile (babel-preset-modern-browsers).
  This was required to enable some ES2017 feature usage, primarily async/await.
- eslint configuration to more closely match Prettier defaults
- a start at de-linting our JS code
- fixed a bunch of bugs, notably:
  - TypedCrossfilter::countFiltered() would return the incorrect count when number of
    dimensions exceeded 32.
  - regraph/reset was fundamentally broken all over the place. For example, none of
    the metadata pickers in the left sidebar would correctly update to match the
    current `world`

Things to note about the new `utils/stateManager` code:

- Universe: this contains support for data that we cache client-side, including:

  - schema
  - obs and var annotations - aka cell and gene metadata
  - var data (expression)
  - obs layout (graph)

  `Universe` module implements REST API specific functions to convert from OTA
  formats to the internal structure (eg, REST 0.1 to internal object format).   It does 
  not yet support the new REST 0.2 API. 

  The primary user of Universe is data loading, and serving as the source of
  data for World.

- `World` - the primary interface for the app, providing a view into the subset of
  information that constitutes the current World. Almost without exception, the
  application components/actions/etc should **only** use World for:

  - obs and var annotations
  - var data (expression)
  - obs layout (graph)

  The data model and interfaces presented by World are essentially identical to
  Universe, except that they only manifest a subset of the data (the current
  contents of the World). World only uses the default Universe interfaces.

- Other state that needs to be kept consistent with "world" is maintained in the 
  reducers, and includes:

  - selection state (crossfilter object)
  - per-cell color name and color RGB

- `kvCache` - a simple LRU cache for expression data, used by both World and
  Universe to maintain recently used var (expression) data

colinmegill commented on client/src/util/stateManager/universe.js:
```diff
@@ -0,0 +1,290 @@
+    this.nVar = this.varAnnotations.length;
+    this.init.initialize = true;
+    this._tryFinalization();
+    return this;
```
Prefer to avoid this pattern, open to discussion

bkmartinjr replied:
do you have a preferred pattern?    The order of the init calls is indeterminate, as they are the result of network calls.    So either the caller of the init functions explicitly calls something like _tryFinalization(), or the init calls do it.  I'm agnostic, so let me know what you have in mind.

colinmegill commented on client/src/util/stateManager/universe.js:
```diff
@@ -0,0 +1,290 @@
+    this.obsLayout = { X: [], Y: [] };
+
+    /*
+    Expression cache for var/gene data, keyed by varIndex.  Loaded
```
This comment could be expanded. The comment within the keyval cache file is quite extensive - I think it's warranted to speak about the motivations for this function here in narrative form, and the role it plays in the overall system

bkmartinjr replied:
I expanded the comment here and in world.js.   Please review and if acceptable, resolve this comment.   I was not sure exactly what aspects you wanted addressed, so I hit the topics I thought would be most confusing.

colinmegill commented on client/src/reducers/universe.js:
```diff
@@ -0,0 +1,65 @@
+TODO: 0.2 reducer
+*/
+
+const UniverseReducerV01 = (state = new Universe("0.1"), action) => {
```
Worth a discussion here. https://redux.js.org/faq/organizingstate#can-i-put-functions-promises-or-other-non-serializable-items-in-my-store-state 

This threw me at first and it took me a minute to find out whether this was something I hadn't done because of my use cases or convention or recommendations or something else. It turned out - see above - plain objects are recommended (and what I would prefer to stick with if possible). I would be interested to discuss converting this to a series of functions and storing the result in redux. If at the end of that conversation this still makes more sense, I'll relent, but I think `return this` is indicating there may be profit / simplicity in a more functional approach.

bkmartinjr replied:
I think you are right - if we want the "pure" (serializable) approach advocated in the FAQ link.   Note that we have the same issue with crossfilter.  While crossfilter is not implemented as an ES6 class, it is not a pure object (it has a prototype with functions).

I need some of your time to understand the right approach.     Universe will be simple to convert to functions & a pure object.     World and crossfilter will be somewhat more complex but possible.

Issues I can see off top of my head:
* Loss of sugar:  all of the `state.world.crossfilter.dosomething()` will become `crossfilterModule.dosomething(state.world.crossfilter)` where `dosomething()` includes the complex getters, setters, etc.
* all embedded objects (crossfilter, KeyValueCache, etc) have to be similarly functionalized, which will increase verbosity.   This affects both modules (stateManager and crossfilter), as they have a lot of embedded non-pure objects.

This will require a significant refactoring of these modules, so I definitely want to understand the requirements.   F2F...

colinmegill commented:
Took a first pass, surfaced several aspects for consideration. Let's discuss those and then I'll take another pass.

bkmartinjr commented on client/src/util/stateManager/universe.js:
```diff
@@ -0,0 +1,290 @@
+/*
+Universe on top of the REST 0.1 interface
+*/
+class UniverseV01 extends UniverseBase {
```
accepted.

bkmartinjr replied:
One more note: our current eslint config requires camelcase.  We should decide if snakecase is allowed or not.

bkmartinjr commented on client/src/actions/index.js:
```diff
@@ -1,101 +1,157 @@
+    if (!thereWasAnError) {
+      dispatch({
+        type: "initial data load complete (universe exists)",
+        universe: getState().universe
```
There are a couple of issues here (most I can clean up).   However, the core issue is how one reducer hands state/info to another reducer during initialization.

In this case:
* universe reducer creates Universe
* controls reducer creates World, which needs to know about Universe

What is the right pattern?  I can see:
* create a single reducer that does both
* the action dispatch the newly created Universe to the controls/world reducer
* middleware links the two

I think it boils down to a basic question: is it reasonable to have state managed by separate reducers linked / dependent in any way?   Or does that linkage/dependency (in this case, world/universe) imply they should be a single reducer?

bkmartinjr replied:
I have another approach that I'll mock up prior to our F2F.

bkmartinjr replied:
I have fixed this by integrating the universe into the controls reducer (it should have been there all along).

bkmartinjr commented on client/src/actions/index.js:
```diff
@@ -1,101 +1,157 @@
-            }
-          }
-        });
+    if (!thereWasAnError) {
```
fixed

bkmartinjr commented:
@colinmegill - at this point I believe I have addressed all of the comments except two:
* the comment about patterns in stateManager/universe.js
* the use of ES6 classses in the redux state

I'm going to await further elaboration on the first.  On the second, I need to do more research and will propose something.

colinmegill commented on client/src/reducers/controls2.js:
```diff
@@ -0,0 +1,333 @@
+      const colorName = new Array(universe.nObs).fill(globals.defaultCellColor);
+      const colorRGB = _.map(colorName, c => parseRGB(c));
+      const categoricalAsBooleansMap = createCategoricalAsBooleansMap(world);
+      const obsCrossfilter = crossfilter(world.obsAnnotations);
```
It seems like we only have one crossfilter, and that will stick. I would prefer the pattern: 

`import Crossfilter from ...`
`const crossfilter = Crossfilter(world.obsAnnotations)`

Uppercase constructor lowercase instance here as well, and that is 'the' crossfilter for the app, and we know to look for that.

bkmartinjr replied:
today we have one, but if we start filtering by var (gene), we would have two.  Hence the naming `obsCrossfilter` - crossfilters on obs annotations.

that said, I'll change it and we can add more clarity when / if we need disambiguation.

bkmartinjr replied:
i assume you want the same change - for the same reason - for obsDimensionMap.  It contains the dimensions on the obs crossfilter.   Will be ambiguous if we ever allow crossfilters on both, but for now it is a moot point.

colinmegill commented:
Overall, this PR seems like a massive improvement and simplification in async data flow, client side data processing, client side state management and the general abstraction patterns the app has implemented to this point in `universe` and `world`. Great work @bkmartinjr!

bkmartinjr commented on client/src/reducers/controls2.js:
```diff
@@ -0,0 +1,333 @@
+// jshint esversion: 6
+
+import _ from "lodash";
+import { World, kvCache } from "../util/stateManager";
```
Yeah, KeyValueCache is simple a utility class.   Can hold anything (it is an Object under the covers).   So I named it by the data that it holds - in this case it caches expression data (or more accurately in our new terminology, responses from /data/var - so `varDataCache`).

if you have a better name, lets fix it.

bkmartinjr replied:
filed as issue #239

bkmartinjr commented on client/src/util/stateManager/universe.js:
```diff
@@ -0,0 +1,252 @@
+Cherry pick from /api/v0.1 response format to make somethign similar
+to the v0.2 schema, which we use for internal interfaces.
+*/
+function OtaRESTv01ToSchema(ota) {
```
will fix.

bkmartinjr commented on client/src/components/graph/graph.js:
```diff
@@ -130,34 +142,31 @@ class Graph extends React.Component {
-        this.props.crossfilter.cells != prevProps.crossfilter.cells
+        selectionUpdate != prevProps.selectionUpdate
       ) {
         if (!this.renderCache.positions)
```
parenthesis or brackets?   

our prettier/eslint config removes any unnecessary parenthesis - we would have to change it to make this preference stick.   and there are a lot of other places we would need to make the same change.

bkmartinjr replied:
filed as issue #240

bkmartinjr commented on client/src/components/continuous/continuous.js:
```diff
@@ -17,18 +17,15 @@ import HistogramBrush from "./histogramBrush";
-    axesHaveBeenDrawn: state.controls.axesHaveBeenDrawn
+    colorAccessor: state.controls2.colorAccessor,
+    colorScale: state.controls2.colorScale,
+    selectionUpdate: _.get(state.controls2, "obsCrossfilter.updateTime", null)
```
Correct, but it is the only way we know when the crossfilter selection is updated.

Because crossfilter is not an immutable POJO, we have to connect it to the components using something that does change.   In this case, the crossfilter has a property called `updateTime` that is guaranteed to be modified *anytime* the crossfilter selection changes.

bkmartinjr commented on client/src/reducers/index.js:
```diff
@@ -1,36 +1,23 @@
-  differential,
-  responsive
+  responsive,
+  controls2,
```
done
